### DISCUSSION TITLE
getting_started/step_by_step/instancing

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#64] [User-contributed notes]
# getting_started/step_by_step/instancing

In the previous part, we saw that a scene is a collection of nodes organized in a tree structure, with a single node as its root. You can split your project into any number of scenes. This feature ...

https://docs.godotengine.org/en/stable/getting_started/step_by_step/instancing.html

**Comment by TheApproStorm** [comment]
Help, whenever i try to open the instancing project the whole thing crashes

**Reply by Calinou** [reply]
Which Godot version are you using and what graphics card model do you have?

The instancing starter uses the Forward+ rendering method, which may not work if your GPU doesn't support Vulkan. You could try opening the project with the `--rendering-method gl_compatibility` [command line argument](https://docs.godotengine.org/en/stable/getting_started/editor/command_line_tutorial.html) as a workaround.

Eventually, all starter projects should be made to use Compatibility (as they don't use advanced graphics features), but this might work for now.

**Reply by v012345** [reply]
use Godot Engine v4.3.stable , I can open the project and run well

**Reply by brianhsu** [reply]
Use Godot 4.3 stable on Linux also works fine for me.

**Reply by rreshrr** [reply]
hi. me helped installing vulkan lib

**Comment by Pascal-CP** [comment]
I think the "Weapon" box above the "Player Ship" box in the shooter game diagram should be called "Player" ?

**Reply by BhavinTheGreat** [reply]
@Pascal-CP its correct as its referring to the weapon that may be put or installed on the ship like a battleship.

**Reply by rvelasquez** [reply]
@BhavinTheGreat Missile is an instance of Enemy, perhaps a more specific name would make it clear like ShipWeapon/Canon/Turret would make it more clear?

**Reply by BhavinTheGreat** [reply]
> @BhavinTheGreat Missile is an instance of Enemy, perhaps a more specific name would make it clear like ShipWeapon/Canon/Turret would make it more clear?

yup that would do too

**Reply by v-kydela** [reply]
But why is a weapon instantiating a weapon? Should it be instantiating a bullet perhaps?

**Comment by das-nayan** [comment]
"To make a resource unique for one instance, right-click on it in the Inspector and click Make Unique in the contextual menu." Make Unique is not showing in the contextual menu for Gravity Scale resource, even though it isn't unique. Please correct me.

**Reply by Calinou** [reply]
**Gravity Scale** is not a resource, but a property of the RigidBody2D node. This is what you should click instead:

![image](https://github.com/user-attachments/assets/957f8a05-fdc8-402e-91fa-628b5a17a4c0)

**Reply by das-nayan** [reply]
Thank you!

**Reply by bill-in-iowa** [reply]
I think it would clear up confusion if you say right click on PhysicsMaterial instead of saying right click on it!!.

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/10032

**Comment by Josias-Cruz** [comment]
When I tried to import the project I just got a black screen with the project tile at the top, 

I'm on Windows 10, and Intel(R) HD Graphics 520

**Comment by JJODY4** [comment]
So if I make a game with multiple playable characters with unique abilities each, i should make a basePlayer and then instantiate? Do I do that for the levels as well?

**Reply by rvelasquez** [reply]
Yes, this is quite common.  In basePlayer you would have shared attributes that all players have in common.  Similar with levels.  However, usually in object oriented languages you just call it Player, not basePlayer.

**Comment by YasinMahdizadeh** [comment]
Hi, Thanks for the great documentation!

In shooter game diagram, many arrows enter the stage scene and tip of arrows form a circle like shape, That was hard to read and it took me some time to figure out that they are actually arrows.

**Comment by stann-co** [comment]
it's not clear from this how to instance from code

**Reply by Calinou** [reply]
The *Scripting languages* page comes after this one, so I think it's by design that instantiating from code is not mentioned.

We could add a quick mention of using `PackedScene.instantiate()`, but it may be confusing for those who are reading the documentation in order and have never used a scripting language before.

**Reply by stann-co** [reply]
I found the article through google, whilst searching for "in code" And it didn't stand out to me in the page overview, maybe it could be added to the name of whatever page it shows up on.
The real fault is probably just on google SEO

**Comment by jakeometer** [comment]
Hello, I don't seem to have "PhysicsMaterial" under Inspector for the ball, can anybody help please?

**Reply by Calinou** [reply]
Did you see [this note above](https://github.com/godotengine/godot-docs-user-notes/discussions/64#discussioncomment-10685128)? You need to create a new PhysicsMaterial resource in the Physics Material property of the RigidBody2D node.

**Reply by chihwahli** [reply]
Please help, there is no text or button to create PhysicsMaterial at all. I opened everything in the inspector of the ball.tscn.
Might be more handy to change the tutorial. Cannot continue.

**Reply by Piralein** [reply]
Does your `ball.tscn` contain a node called `Ball` with the type `RigidBody2D`?
You can mouseover the `Ball` node in the `scene` tree to see its type.

If you select this node it should show a `RigidBody2D` group inside the `inspector`, it should contain the PhysicsMaterial.
If not, what properties are displayed and how is the uppermost group called?

**Reply by azrael4h** [reply]
There isn't a type called RigidBody2D in the inspector. There's a CollisionShape2D, Node2D, CanvasItem, and Node shown in the inspector when selecting the Ball.tscn in the file system. Furthermore, there is no means to actually add PhysicsMaterial; rightclicking on any of those just brings up a documentation drop box, and left clicking just selects them in the inspector. 

None of those have the RigidBody2D under them either. It does not exist under the Ball.tscm node. Godot 4.3, just downloaded both the updated Godot and the asset files today.

**Reply by azrael4h** [reply]
Selecting Main.tscm does have the RigidBody2D, but you cannot actually edit the properties of the PhysicsMaterial. The Bounce, Friction, etc... is greyed out and inaccessible.

**Reply by Grimslip** [reply]
I struggled with this for a while, until I realised that "Ball" was not selected in the Scene. I had selected the ball by clicking on it in the viewport, which in turn selected the CollisionShape2D node. Once I clicked on Ball in the Scene panel, all was well.

**Comment by mirkojotic** [comment]
> The arrows tell you which scene owns which.

is this per chance meant to mean:

> The arrows point to the owner of the scene.

**Reply by Calinou** [reply]
Yes :slightly_smiling_face: 

Pull request opened:

- https://github.com/godotengine/godot-docs/pull/10487

**Comment by UkaSystems** [comment]
Why not add a simple code examples for instancing?

**Reply by Calinou** [reply]
This tutorial comes before the Scripting languages tutorial, so it would be counter-intuitive if code was mentioned before you've learned about it.

However, I've added a link to the other tutorial that specifically covers instancing scenes through code:

- https://github.com/godotengine/godot-docs/pull/10622

**Comment by danieljacobsen-ai** [comment]
"The arrows tell you which scene owns which."

What does that even mean? Owns = is the abstract scene resource for? There is no textual description here of how one scene can "own" another. 

It seems that some arrows mean "contained in", for example a table scene instances is placed insides a room scene instance, and therefore there is an arrow from "table" to room."

**Comment by rdepew** [comment]
The instancing_starter.zip doesn't work with Godot v4.4.

**Reply by Calinou** [reply]
https://github.com/godotengine/godot-docs-project-starters/pull/12 updates all starter projects for 4.4, but it should work already if you click **Upgrade & Restart** in the UID dialog that shows up upon editing the project.

**Reply by knightmb** [reply]
The example instancing_starter still shows up as 4.2 in v4.4 when I downloaded it. It can be a little confusing at first, but the directions to just upgrade and restart seem to work fine. I had to scroll all the way down here to find it because I didn't want to get off track from the instructions.

**Comment by anugrhaswi** [comment]
Help!! im really frustrated as i cannot find the bounce attribute under RigidBody2D????

**Reply by Calinou** [reply]
The Bounce property is part of the PhysicsMaterial resource, which you create in the Physics Material property.

**Reply by vaishnavgade** [reply]
I was able to access the Bounce Attribute after selecting PhysicsMaterial Dropdown and clicking Edit. Hope that helps.

**Comment by Talketsu** [comment]
How come the HUD is instanced from the Market? Or is the idea that the HUD in this theoretical shooter would contain an instances of a Market scene? 

Also is the idea with the enemy instancing on a Missile that Enemy is shooting a Missile and thus the Missile comes from the Enemy? In that case why is the Enemy instanced on the Missile? 

Is the lone "weapon" instanced by code meant to be a random weapon on the stage to pick-up? 

There's a lot of assumptions we need to make about the shooter in place to understand this diagram, it would be good to maybe bring up an example game and show how the design language ties to that in specific, cause there's a couple things in the diagram as is that I find difficult to grasp and makes the concept seem potentially more complicated than I understand it to be.

**Reply by BitterCandies** [reply]
The arrows point TOWARDS the parent node, as I understand it. Market comes FROM HUD, not HUD from market.

The Market is instanced from the HUD. I imagine the market is a special type of HUD.

The missile is instanced from enemy, presumably since the missile can have enemy traits like HP (technically- if you can destroy the missile while its in the air). So the Missile is a special type of enemy.

Yes, presumably the weapon instanced from weapon refers to things like different weapon types? that's how I understand it at least. Like if it's a shooter, the weapon instance of weapon would be different things like Shotgun vs Rifle.

**Reply by Sey97** [reply]
Applying this same analogy to the open world architecture example becomes more confusing and 'unrealistic'. If the arrow points to the parent node does it mean 'Player' is the parent node since 'HUD' is pointing towards it? Is 'HUD' created by the player? 
PS: I'm new to game development and trying to wrap my head around this :)

**Comment by mleaft** [comment]
Was following along with this tutorial on the 4.6 RC2 and the 'make unique' option was greyed out for right-clicking the Physics Material. Going back to 4.5 seemed to solve it.

**Reply by Calinou** [reply]
This is a consequence of this change in 4.6:

- https://github.com/godotengine/godot/pull/109458

You can't use **Make Unique** anymore on resources that are already unique and are built-in to the scene. For it to be available, the resource must be used in at least 2 different locations *or* be an external resource saved to disk.

That said, there are some known bugs and UX quirks with this new implementation. See https://github.com/godotengine/godot/issues/114906 and https://github.com/godotengine/godot/pull/113869 which aims to address this.

**Comment by WielorybGrotolaz** [comment]
I don’t see a single script in the downloaded project, yet balls appear when I click the left mouse button. How is this implemented?

**Reply by Noahzette** [reply]
There is indeed code in the project; you must have missed it, it happens.
<img width="1290" height="388" alt="Capture d&#39;écran 2026-05-18 191522" src="https://github.com/user-attachments/assets/75a3f4f3-5abb-4208-b792-849c02b5b8ab" />
<img width="406" height="434" alt="Capture d&#39;écran 2026-05-18 191417" src="https://github.com/user-attachments/assets/96fc263c-4eb1-4a2f-927a-871d77896455" />

**Reply by WielorybGrotolaz** [reply]
You’re right. I thought that when you open the “Script” tab, it shows all the scripts used in the project.
